Data/ByteString.hs

```diff
@@ -276,27 +277,22 @@ import GHC.Word hiding (Word8)
 
 -- | /O(1)/ Convert a 'Word8' into a 'ByteString'
 singleton :: Word8 -> ByteString
-singleton c = unsafeCreate 1 $ \p -> poke p c
+-- Taking a slice of some static data rather than allocating a new
+-- buffer for each call is nice for several reasons. Since it doesn't
+-- involve any side effects hidden in a 'GHC.Magic.runRW#' call, it
+-- can be simplified to a constructor application. This may enable GHC
+-- to perform further optimizations after inlining, and also causes a
+-- fresh singleton to take only 4 words of heap space instead of 9.
+-- (The buffer object itself would take up 3 words: header, size, and
+-- 1 word of content. The ForeignPtrContents object used to keep the
+-- buffer alive would need two more.)
+singleton c = unsafeTake 1 $ unsafeDrop (fromIntegral c) allBytes
 {-# INLINE singleton #-}
 
---
--- XXX The use of unsafePerformIO in allocating functions (unsafeCreate) is critical!
---
--- Otherwise:
---
---  singleton 255 `compare` singleton 127
---
--- is compiled to:
---
---  case mallocByteString 2 of
---      ForeignPtr f internals ->
---           case writeWord8OffAddr# f 0 255 of _ ->
---           case writeWord8OffAddr# f 0 127 of _ ->
---           case eqAddr# f f of
---                  False -> case compare (GHC.Prim.plusAddr# f 0)
---                                        (GHC.Prim.plusAddr# f 0)
---
---
```

Comment:
The example is outdated now, but are there reasons for removing the entire note regarding `unsafePerformIO`? Admittedly the explanation isn't very good.

Reply:
This incident is still included in the `accursedUnutterablePerformIO` [trail of destruction](https://github.com/haskell/bytestring/blob/0badf9754bae22b37431bf66cefab29ce8e87b4f/Data/ByteString/Internal.hs#L802).

Reply:
Alright. :)